node/src/types.rs

```diff
@@ -48,3 +52,151 @@ pub type NodeRng = ChaCha20Rng;
 /// The RNG used throughout the node for testing.
 #[cfg(test)]
 pub type NodeRng = crate::testing::TestRng;
+
+/// An in-memory object that can possibly be shared with user parts of the system.
+///
+/// In general, this should only be used for immutable, content-addressed objects.
+///
+/// This type exists solely to switch between `Box` and `Arc` based behavior, future updates should
+/// deprecate this in favor of using `Arc`s directly or turning `LoadedItem` into a newtype.
+#[derive(Clone, DataSize, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub enum SharedObject<T> {
+    /// An owned copy of the object.
+    Owned(Box<T>),
+    /// A shared copy of the object.
+    Shared(Arc<T>),
+}
+
+impl<T> Deref for SharedObject<T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        match self {
+            SharedObject::Owned(obj) => &*obj,
+            SharedObject::Shared(shared) => &*shared,
+        }
+    }
+}
+
+impl<T> AsRef<[u8]> for SharedObject<T>
+where
+    T: AsRef<[u8]>,
+{
+    fn as_ref(&self) -> &[u8] {
+        match self {
+            SharedObject::Owned(obj) => <T as AsRef<[u8]>>::as_ref(obj),
+            SharedObject::Shared(shared) => <T as AsRef<[u8]>>::as_ref(shared),
+        }
+    }
+}
+
+impl<T> SharedObject<T> {
+    /// Creates a new owned instance of the object.
+    #[inline]
+    pub fn owned(inner: T) -> Self {
+        SharedObject::Owned(Box::new(inner))
+    }
+
+    /// Creates a new shared instance of the object.
+    pub fn shared(inner: Arc<T>) -> Self {
+        SharedObject::Shared(inner)
+    }
+}
+
+impl<T> Display for SharedObject<T>
+where
+    T: Display,
+{
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            SharedObject::Owned(inner) => inner.fmt(f),
+            SharedObject::Shared(inner) => inner.fmt(f),
+        }
+    }
+}
+
+impl<T> Serialize for SharedObject<T>
+where
+    T: Serialize,
+{
+    #[inline]
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            SharedObject::Owned(inner) => inner.serialize(serializer),
+            SharedObject::Shared(shared) => shared.serialize(serializer),
+        }
+    }
+}
+
+impl<'de, T> Deserialize<'de> for SharedObject<T>
+where
+    T: Deserialize<'de>,
+{
+    #[inline]
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        T::deserialize(deserializer).map(SharedObject::owned)
```

Comment:
There's no way of storing it in the cache at this point - we do not have access to the cache here. So any sharing would just be an `Arc` with a single reference.

There is merit to using `shared` regardless, as we would be able to avoid a copy if we update the code to reuse the `Arc`. But currently this has no effect, since the contained data is straight up deserialized and thrown away.

So the gain is not as big as it could be right now and I err'd on the side of keeping the old behavior for the sake of not making any more behavioral changes than necessary.

Reply:
Thanks - makes sense.